### PR TITLE
Define functions on scalars and vectorise them

### LINKED ISSUE #281: Automate aggregation of individual measures
### Current and desired situation

Currently, the code contains several functions that aggregate individual transfers or taxes on the tax unit or household level (e.g. `kindergeld_m_hh`). 

The function definitions all look basically the same.

Would be pretty to somehow define these functions without repeating code.

### Proposed implementation

not sure yet

### Considered alternatives

### PR DESCRIPTION
**Pull request by ChristianZimpelmann**
### What problem do you want to solve?

- Make functions easier to read and understand
- Make code much faster

### Todo

- [x] Implement functions to aggregate values on hh or tu level
- [x] Rewrite all functions such that they are defined on scalars
- [x] Implement vectorization (for numpy ~~, numba,~~ and jax)
- [x] Rewrite `interface.py` such that all columns are on individual level
- [x] If it is a relevant contribution, put it in `CHANGES.rst`.

Closes #281

### CODE REVIEW

**Comment by hmgaudecker** [comment]
Tests still fail due to:

```pytb
gettsim/aggregation_numpy.py:56: in grouped_min
    out_on_hh = npg.aggregate(group_id, column, func="min", fill_value=fill_value)
/home/hmg/miniconda3/envs/gettsim/lib/python3.9/site-packages/numpy_groupies/aggregate_numpy.py:291: in aggregate
    return _aggregate_base(group_idx, a, size=size, fill_value=fill_value,
/home/hmg/miniconda3/envs/gettsim/lib/python3.9/site-packages/numpy_groupies/aggregate_numpy.py:280: in _aggregate_base
    ret = func(group_idx, a, flat_size, fill_value=fill_value, dtype=dtype,
/home/hmg/miniconda3/envs/gettsim/lib/python3.9/site-packages/numpy_groupies/aggregate_numpy.py:82: in _min
    else np.finfo(a.dtype).max
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'numpy.finfo'>, dtype = <class 'numpy.datetime64'>

    def __new__(cls, dtype):
        try:
            dtype = numeric.dtype(dtype)
        except TypeError:
            # In case a float instance was given
            dtype = numeric.dtype(type(dtype))
    
        obj = cls._finfo_cache.get(dtype, None)
        if obj is not None:
            return obj
        dtypes = [dtype]
        newdtype = numeric.obj2sctype(dtype)
        if newdtype is not dtype:
            dtypes.append(newdtype)
            dtype = newdtype
        if not issubclass(dtype, numeric.inexact):
>           raise ValueError("data type %r not inexact" % (dtype))
E           ValueError: data type <class 'numpy.datetime64'> not inexact

/home/hmg/miniconda3/envs/gettsim/lib/python3.9/site-packages/numpy/core/getlimits.py:473: ValueError
```

Should I just ignore? Should we skip those?

**Comment by ChristianZimpelmann** [comment]
The PR is now ready for review. A few comments:

- I haven't implemented user-provided aggregation functions, yet. Will do so tomorrow.
- Also missing is a final update of the tutorials.
- I added a `_hh` suffix to a few column names which are expected to be constant within households.
- There is now a separate test workflow that runs jax and uses the environment `environment_linux_jax.yml`. While adding those tests, I dropped support for python 3.7.

File: gettsim/social_insurance_contributions/beitr_bemess_grenzen.py [hunk 1]
```diff
@@ -45,20 +39,19 @@ def _ges_krankenv_beitr_bemess_grenze_m(
...
-            True: soz_vers_beitr_params["beitr_bemess_grenze_m"]["ges_krankenv"]["ost"],
-            False: soz_vers_beitr_params["beitr_bemess_grenze_m"]["ges_krankenv"][
-                "west"
-            ],
-        }
-    )
-    return out.astype(float)
+    params = soz_vers_beitr_params["beitr_bemess_grenze_m"]["ges_krankenv"]
+
+    if wohnort_ost:
```
Comment (hmgaudecker) [inline gettsim/social_insurance_contributions/beitr_bemess_grenzen.py:44]:
```suggestion
    out = params["ost"] if wohnort_ost else params["west"]
```
is much more readable (that's why the construct was created fairly recently). No need to change everywhere, but this particular instance appears so often that it seems worth a search & replace to me.

File: gettsim/social_insurance_contributions/eink_grenzen.py [hunk 2]
```diff
@@ -15,23 +11,16 @@ def mini_job_grenze(wohnort_ost: BoolSeries, soz_vers_beitr_params: dict):
...
-        {
-            True: soz_vers_beitr_params["geringfügige_eink_grenzen_m"]["mini_job"][
-                "ost"
-            ],
-            False: soz_vers_beitr_params["geringfügige_eink_grenzen_m"]["mini_job"][
-                "west"
-            ],
-        }
-    )
+    mini_job_params = soz_vers_beitr_params["geringfügige_eink_grenzen_m"]["mini_job"]
```
Comment (hmgaudecker) [inline gettsim/social_insurance_contributions/eink_grenzen.py:14]:
```suggestion
    params = soz_vers_beitr_params["geringfügige_eink_grenzen_m"]["mini_job"]
```
As above. I like this shortening, but let's be consistent with it.

File: gettsim/social_insurance_contributions/eink_grenzen.py [hunk 3]
```diff
@@ -76,16 +63,17 @@ def in_gleitzone(
 
     Returns
     -------
-    BoolSeries indicating individual's income is in midi-job range.
+    Boolean indicating individual's income is in midi-job range.
```
Comment (hmgaudecker) [inline gettsim/social_insurance_contributions/eink_grenzen.py:66]:
```suggestion
    Whether individual's income is in midi-job range.
```

File: gettsim/social_insurance_contributions/eink_grenzen.py [hunk 4]
```diff
@@ -48,16 +37,14 @@ def geringfügig_beschäftigt(
 
     Returns
     -------
-    BoolSeries indicating if person earns less than marginal employment threshold.
+    Boolean indicating if person earns less than marginal employment threshold.
```
Comment (hmgaudecker) [inline gettsim/social_insurance_contributions/eink_grenzen.py:40]:
```suggestion
    Whether person earns less than marginal employment threshold.
```

File: gettsim/social_insurance_contributions/eink_grenzen.py [hunk 5]
```diff
@@ -76,16 +63,17 @@ def in_gleitzone(
...
+    ) & (not geringfügig_beschäftigt)
+    return out
 
 
 def midi_job_bemessungsentgelt_m(
-    bruttolohn_m: FloatSeries, in_gleitzone: BoolSeries, soz_vers_beitr_params: dict,
-) -> FloatSeries:
+    bruttolohn_m: float, soz_vers_beitr_params: dict,
+) -> float:
     """Select income subject to social insurance contributions for midi job.
```
Comment (hmgaudecker) [inline gettsim/social_insurance_contributions/eink_grenzen.py:77]:
```suggestion
    """Income subject to social insurance contributions for midi job.
```

File: gettsim/social_insurance_contributions/eink_grenzen.py [hunk 6]
```diff
@@ -98,15 +86,13 @@ def midi_job_bemessungsentgelt_m(
...
-    in_gleitzone
-        See :func:`in_gleitzone`.
     soz_vers_beitr_params
         See params documentation :ref:`soz_vers_beitr_params <soz_vers_beitr_params>`.
 
 
     Returns
     -------
-    FloatSeries with the income subject to social insurance contributions for midi job.
+    float with the income subject to social insurance contributions for midi job.
```
Comment (hmgaudecker) [inline gettsim/social_insurance_contributions/eink_grenzen.py:95]: Do we really need to repeat the dtype?

File: gettsim/social_insurance_contributions/ges_pflegev.py [hunk 7]
```diff
@@ -1,13 +1,6 @@
...
-from gettsim.typing import IntSeries
-
-
 def ges_pflegev_zusatz_kinderlos(
-    hat_kinder: BoolSeries, alter: IntSeries, soz_vers_beitr_params: dict,
-) -> BoolSeries:
+    hat_kinder: bool, alter: int, soz_vers_beitr_params: dict,
+) -> bool:
     """
     Create boolean Series indicating addtional care insurance contribution for
```
Comment (hmgaudecker) [inline gettsim/social_insurance_contributions/ges_pflegev.py:5]:
```suggestion
    Whether additional care insurance contribution for childless individuals applies.
```

File: gettsim/social_insurance_contributions/ges_pflegev.py [hunk 8]
```diff
@@ -1,13 +1,6 @@
...
     childless individuals.
```
Comment (hmgaudecker) [inline gettsim/social_insurance_contributions/ges_pflegev.py:6]:
```suggestion
```

File: gettsim/social_insurance_contributions/ges_pflegev.py [hunk 9]
```diff
@@ -199,14 +198,15 @@ def _ges_pflegev_beitr_midi_job_m_m(
         * soz_vers_beitr_params["soz_vers_beitr"]["ges_pflegev"]["standard"]
     )
 
-    zusatz_kinderlos = (
-        midi_job_bemessungsentgelt_m.loc[ges_pflegev_zusatz_kinderlos]
-        * soz_vers_beitr_params["soz_vers_beitr"]["ges_pflegev"]["zusatz_kinderlos"]
-    )
+    # Add additional contribution for childless individuals
+    if ges_pflegev_zusatz_kinderlos:
+        ges_beitr_midi_job_m += (
```
Comment (hmgaudecker) [inline gettsim/social_insurance_contributions/ges_pflegev.py:203]:
```suggestion
        gesamtbeitrag_midi_job_m += (
```

File: gettsim/taxes/zu_verst_eink/eink.py [hunk 10]
```diff
@@ -190,65 +97,12 @@ def eink_rente_zu_verst(
...
-
 def sum_eink_ohne_kapital(
-    eink_selbst: FloatSeries,
-    eink_abhängig_beschäftigt: FloatSeries,
-    eink_vermietung: FloatSeries,
-    eink_rente_zu_verst: FloatSeries,
-) -> FloatSeries:
+    eink_selbst: float,
+    eink_abhängig_beschäftigt: float,
+    vermiet_eink: float,
```
Comment (hmgaudecker) [inline gettsim/taxes/zu_verst_eink/eink.py:103]: Did we decide to turn this around? I see no reason to change the name (but might have suggested it previously, of course), especially since the official thing is "Einkommen aus Vermietung und Verpachtung", if I am not mistaken.

File: gettsim/taxes/zu_verst_eink/freibeträge.py [hunk 11]
```diff
@@ -119,22 +121,22 @@ def eink_st_altersfreib(
     -------
 
     """
-    out = bruttolohn_m * 0
     agelimit = eink_st_abzüge_params["altersentlastungsbetrag_altersgrenze"]
```
Comment (hmgaudecker) [inline gettsim/taxes/zu_verst_eink/freibeträge.py:124]:
```suggestion
    altersgrenze = eink_st_abzüge_params["altersentlastungsbetrag_altersgrenze"]
```
No need to translate.

File: gettsim/tests/test_tax_transfer.py [hunk 12]
```diff
@@ -14,7 +14,7 @@
 from gettsim.typing import check_if_series_has_internal_type
 
 
-YEARS = [2019, 2020, 2021]
+YEARS = [2019]
```
Comment (hmgaudecker) [inline gettsim/tests/test_tax_transfer.py:17]: Why?

File: gettsim/tests/test_unterhalt_vorschuss.py [hunk 13]
```diff
@@ -1,5 +1,3 @@
-import itertools
```
Comment (hmgaudecker) [inline gettsim/tests/test_unterhalt_vorschuss.py:1]: Filename needs updating.

File: gettsim/transfers/benefit_checks/vermögens_checks.py [hunk 14]
```diff
@@ -1,22 +1,14 @@
...
-
-from gettsim.typing import BoolSeries
-from gettsim.typing import FloatSeries
-from gettsim.typing import IntSeries
-
-
 def _kinderzuschl_nach_vermög_check_m_hh(
-    _kinderzuschl_vor_vermög_check_m_hh: FloatSeries,
-    vermögen_hh: FloatSeries,
+    kinderzuschl_vorläufig_m_hh: float,
```
Comment (hmgaudecker) [inline gettsim/transfers/benefit_checks/vermögens_checks.py:2]: I am pretty sure we explicitly changed that in the big renaming thing. Why back to `vorläufig`? It is ambiguous what is meant by that.

**Review by hmgaudecker — COMMENTED** [review]
Looks great, thanks! Mostly minor comments, not sure about some renamings.

*Note to self: continue at `elterngeld.py`

File: gettsim/transfers/kindergeld.py [hunk 15]
```diff
@@ -1,9 +1,16 @@
...
+        "source_col": "kindergeld_anspruch",
+        "aggr": "sum",
+    },
+}
+
+
+def kindergeld_m_bis_1996(kindergeld_basis_m: float) -> float:
     """Kindergeld calculation until 1996.
 
     Until 1996 individuals could claim child allowance and recieve child benefit.
```
Comment (hmgaudecker) [inline gettsim/transfers/kindergeld.py:16]:
```suggestion
    Until 1996 individuals could claim Kinderfreibetrag and receive Kindergeld at the same time.
```

File: gettsim/transfers/kindergeld.py [hunk 16]
```diff
@@ -40,104 +45,50 @@ def kindergeld_m_ab_1997(
...
-    hh_id
-        See basic input variable :ref:`hh_id <hh_id>`.
-
-    Returns
-    -------
-
-    """
-    return kindergeld_m.groupby(hh_id).sum()
-
+    if kinderfreib_günstiger_tu:
```
Comment (hmgaudecker) [inline gettsim/transfers/kindergeld.py:48]: out = 0 if kinderfreib_günstiger_tu else kindergeld_basis_m

File: gettsim/transfers/kindergeld.py [hunk 17]
```diff
@@ -156,12 +107,10 @@ def kindergeld_anspruch_nach_stunden(
...
     Returns
     -------
-    BoolSeries indiciating kindergeld eligibility.
+    Boolean indiciating kindergeld eligibility.
     """
-    out = alter < kindergeld_params["höchstalter"]["ohne_bedingungen"]
-    out = out | (
-        (kindergeld_params["höchstalter"]["ohne_bedingungen"] <= alter)
-        & (alter <= kindergeld_params["höchstalter"]["mit_bedingungen"])
+    out = (alter < kindergeld_params["höchstalter"]["ohne_bedingungen"]) | (
```
Comment (hmgaudecker) [inline gettsim/transfers/kindergeld.py:112]: Why changed from `<=` to `<` ?, also line below ?  Former seems to make more sense. Pls always test these edge cases!

File: gettsim/transfers/kindergeld.py [hunk 18]
```diff
@@ -196,57 +142,10 @@ def kindergeld_anspruch_nach_lohn(
     -------
 
     """
-    out = alter < kindergeld_params["höchstalter"]["ohne_bedingungen"]
-    out = out | (
-        (kindergeld_params["höchstalter"]["ohne_bedingungen"] <= alter)
-        & (alter <= kindergeld_params["höchstalter"]["mit_bedingungen"])
+    out = (alter < kindergeld_params["höchstalter"]["ohne_bedingungen"]) | (
```
Comment (hmgaudecker) [inline gettsim/transfers/kindergeld.py:145]: Same as before. Why suddently a strict equals ?

File: gettsim/transfers/kinderzuschl/kinderzuschl.py [hunk 19]
```diff
@@ -49,97 +44,89 @@ def kinderzuschl_m_hh(
     -------
 
     """
-    out = _kinderzuschl_nach_vermög_check_m_hh.copy()
-    cond = (~kinderzuschl_vorrang_hh & ~wohngeld_kinderzuschl_vorrang_hh) | (
+    if ((not kinderzuschl_vorrang_hh) & (not wohngeld_kinderzuschl_vorrang_hh)) | (
         anz_rentner_hh > 0
```
Comment (hmgaudecker) [inline gettsim/transfers/kinderzuschl/kinderzuschl.py:48]: Is `anz_rentner_hh > 0` in the law or some simplification? If the latter, it might be useful to make a note for that point in time when we'll try to disentangle `grunds_im_alter` and `arbeitsl_geld_2` better.

File: gettsim/transfers/kinderzuschl/kinderzuschl.py [hunk 20]
```diff
@@ -49,97 +44,89 @@ def kinderzuschl_m_hh(
...
+
+def kinderzuschl_vorläufig_m_hh_bis_06_2019(
+    arbeitsl_geld_2_brutto_eink_m_hh: float,
+    kinderzuschl_eink_min_m_hh: float,
+    kinderzuschl_eink_max_m_hh: float,
+    arbeitsl_geld_2_eink_m_hh: float,
+    kinderzuschl_kindereink_abzug_m_hh: float,
+    kinderzuschl_eink_anrechn_m_hh: float,
+) -> float:
+    """Calculate preliminary child benefit since 2005 until 06/2019.
```
Comment (hmgaudecker) [inline gettsim/transfers/kinderzuschl/kinderzuschl.py:100]:
```suggestion
    """Calculate Kinderzuschlag since 2005 until 06/2019. Whether Kinderzuschlag or Arbeitslosengeld 2 applies will be checked later.
```

File: gettsim/transfers/kinderzuschl/kinderzuschl.py [hunk 21]
```diff
@@ -49,97 +44,89 @@ def kinderzuschl_m_hh(
...
-    kinderzuschl_vorläufig_m: FloatSeries, hh_id: IntSeries
-) -> FloatSeries:
-    """Aggregate preliminary child benefit on household level.
+def kinderzuschl_vorläufig_m_hh_ab_07_2019(
+    arbeitsl_geld_2_brutto_eink_m_hh: float,
+    kinderzuschl_eink_min_m_hh: float,
+    kinderzuschl_kindereink_abzug_m_hh: float,
+    kinderzuschl_eink_anrechn_m_hh: float,
+) -> float:
+    """Calculate preliminary child benefit since 07/2019.
```
Comment (hmgaudecker) [inline gettsim/transfers/kinderzuschl/kinderzuschl.py:63]:
```suggestion
    """Calculate Kinderzuschlag since 07/2019. Whether Kinderzuschlag or Arbeitslosengeld 2 applies will be checked later.
```

File: gettsim/transfers/kinderzuschl/kost_unterk.py [hunk 22]
```diff
@@ -103,56 +84,25 @@ def _kinderzuschl_wohnbedarf_eltern_anteil(
     -------
 
     """
-    kinder_in_tu = tu_id.replace(anz_kinder_tu)
-    erwachsene_in_tu = tu_id.replace(anz_erwachsene_tu)
-    conditions = []
-    choices = []
     ex_min = kinderzuschl_params["exmin"]
```
Comment (hmgaudecker) [inline gettsim/transfers/kinderzuschl/kost_unterk.py:87]:
```suggestion
    ex_min = kinderzuschl_params["existenzminimum"]
```

No need to abbreviate (and use two different abbreviations)

File: gettsim/transfers/rente.py [hunk 23]
```diff
@@ -67,20 +40,22 @@ def ges_rente_nach_grundr_m(
     -------
 
     """
-    out = ges_rente_vor_grundr_m + grundr_zuschlag_m
-
     # Return 0 if person not yet retired
-    out.loc[~rentner] = 0
+    if rentner:
```
Comment (hmgaudecker) [inline gettsim/transfers/rente.py:44]: one-liner

File: gettsim/transfers/rente.py [hunk 24]
```diff
@@ -109,14 +84,16 @@ def ges_rente_vor_grundr_m(
 
     """
 
-    out = entgeltp_update * ges_rente_zugangsfaktor * rentenwert
+    # Return 0 if person not yet retired
+    if rentner:
```
Comment (hmgaudecker) [inline gettsim/transfers/rente.py:88]: one-lineer

File: gettsim/transfers/rente.py [hunk 25]
```diff
@@ -130,16 +107,17 @@ def rentenwert(wohnort_ost: BoolSeries, ges_rente_params: dict) -> FloatSeries:
...
-    out = wohnort_ost.replace(
-        {
-            True: ges_rente_params["rentenwert"]["ost"],
-            False: ges_rente_params["rentenwert"]["west"],
-        }
-    ).astype(float)
-    return out
+    params = ges_rente_params["rentenwert"]
+
+    if wohnort_ost:
```
Comment (hmgaudecker) [inline gettsim/transfers/rente.py:112]: one-liner

File: gettsim/transfers/rente.py [hunk 26]
```diff
@@ -153,18 +131,17 @@ def rentenwert_vorjahr(wohnort_ost: BoolSeries, ges_rente_params: dict) -> Float
...
-    out = wohnort_ost.replace(
-        {
-            True: ges_rente_params["rentenwert_vorjahr"]["ost"],
-            False: ges_rente_params["rentenwert_vorjahr"]["west"],
-        }
-    ).astype(float)
-    return out
+    params = ges_rente_params["rentenwert_vorjahr"]
+
+    if wohnort_ost:
```
Comment (hmgaudecker) [inline gettsim/transfers/rente.py:136]: one-liner

File: gettsim/transfers/rente.py [hunk 27]
```diff
@@ -218,32 +196,35 @@ def entgeltp_update_lohn(
     """
 
     # Scale bruttolohn up if earned in eastern Germany
-    bruttolohn_scaled_east = bruttolohn_m
-    bruttolohn_scaled_east.loc[wohnort_ost] = (
-        bruttolohn_scaled_east.loc[wohnort_ost]
-        * ges_rente_params["umrechnung_entgeltp_beitrittsgebiet"]
-    )
+    if wohnort_ost:
```
Comment (hmgaudecker) [inline gettsim/transfers/rente.py:199]: Oh, maybe make a note/issue: I thought this parameters only applies to what was earned during GDR times, now current stuff ?

File: gettsim/transfers/wohngeld.py [hunk 28]
```diff
@@ -365,54 +331,49 @@ def wohngeld_miete_m_bis_2008(
...
-        else wohngeld_params["max_miete"][max_definierte_hh_größe][constr_year][ms]
-        + wohngeld_params["max_miete"]["jede_weitere_person"][constr_year][ms]
-        * (hh_größe - max_definierte_hh_größe)
-        for hh_größe, constr_year, ms in zip(
-            haushaltsgröße, constr_year_category, mietstufe
+
+    constr_year = list(params_max_miete[1])[selected_bin_index]
+
+    # Calc maximal considered rent
+    max_definierte_hh_größe = max(i for i in params_max_miete if isinstance(i, int))
```
Comment (hmgaudecker) [inline gettsim/transfers/wohngeld.py:345]: Please make note / issue to do more pre-processing of yaml contents. We don't want to repeat this a gazillion times, also not very robust.

**Review by hmgaudecker — COMMENTED** [review]
Great! Just a couple of simplifications, notes, very few substantive points. Thanks!

File: .github/workflows/continuous-integration-workflow.yaml [hunk 29]
```diff
@@ -16,7 +16,7 @@ jobs:
       fail-fast: false
       matrix:
         os: ["ubuntu-latest", "macos-latest", "windows-latest"]
-        python-version: ["3.7", "3.8", "3.9"]
+        python-version: ["3.8", "3.9"]
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply .github/workflows/continuous-integration-workflow.yaml:19]: Good point. Will include 3.10, too.

File: environment_linux_jax.yml [hunk 30]
```diff
@@ -0,0 +1,31 @@
+name: gettsim
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply environment_linux_jax.yml:1]: For the aggregation tests, it was still relevant. But I agree that we don't need them at the moment.

[the remark this answers is not in the document]
Comment (hmgaudecker) [reply environment_linux_jax.yml:1]: Your reply does not seem to fit the comment here. Or am I missing something?

File: .github/workflows/continuous-integration-workflow.yaml [hunk 31]
```diff
@@ -29,17 +29,34 @@ jobs:
...
 
       - name: Run pre-commits.
-        if: runner.os == 'Linux' && matrix.python-version == '3.8'
+        if: runner.os == 'Linux' && matrix.python-version == '3.9'
         shell: bash -l {0}
         run: |
           pre-commit install -f --install-hooks
           pre-commit run --all-files
+
+  run-jax-tests:
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply .github/workflows/continuous-integration-workflow.yaml:41]: Should belong here: For the aggregation tests, it was still relevant. But I agree that we don't need them at the moment.

File: environment_linux_jax.yml (hunk 30 above)
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply environment_linux_jax.yml:1]: Yes! Copied it to the correct comment

**Comment by hmgaudecker** [comment]
Also still needed: Search for "Series" (uppercase) and replace appropriately at least in the GETTSIM functions (not necessary to go in detail through `interface.py` right now). 

There are quite some instances in the docs that were not using "FloatSeries" and friends before.

File: gettsim/aggregation_numpy.py [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+
+    # Expand to indididual level
+    out = out_on_hh[group_id]
+    return out
+
+
+def grouped_max(column, group_id):
+    fail_if_dtype_of_group_id_not_int(group_id, agg_func="max")
+    fail_if_dtype_not_numeric_or_datetime(column, agg_func="max")
+    fill_value = (
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply gettsim/aggregation_numpy.py:40]: I added the following:
```
    # numpy_groupies expects a `fill_value` which is assigned to groups which are not
    # present in the data (e.g. if the data only contains the group_id 0 and 4, group
    # 1-3 would get this value). This isn't relevant here because grouped_max returns
    # the values on the individual level. Nevertheless, we need to make sure that the
    # fill_value is convertible to the dtype of `column`.

File: gettsim/demographic_vars.py [hunk 33]
```diff
@@ -27,301 +41,154 @@ def anz_kinder_bis_17_hh(
...
+
     """
     return anz_erwachsene_hh == anz_rentner_hh
 
 
-def date_of_birth(
-    geburtsjahr: IntSeries, geburtsmonat: IntSeries, geburtstag: IntSeries
-) -> DateTimeSeries:
+def geburtstermin(
+    geburtsjahr: int, geburtsmonat: int, geburtstag: int
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply gettsim/demographic_vars.py:193]: That would be a good change. I will open an issue. Let's stick with `geburtstag` until then.

File: gettsim/taxes/zu_verst_eink/eink.py (hunk 10 above)
Comment (ChristianZimpelmann) [reply gettsim/taxes/zu_verst_eink/eink.py:103]: I also don't remember.. changed it to `eink_vermiet` now

Comment (hmgaudecker) [reply gettsim/taxes/zu_verst_eink/eink.py:103]: Why not leave it? We decided on it that way for a reason...

File: gettsim/tests/test_tax_transfer.py (hunk 12 above)
Comment (ChristianZimpelmann) [reply gettsim/tests/test_tax_transfer.py:17]: The current test data only contains test cases for 2019. (No clue, why the code looped over 2019-2021 before)

File: gettsim/transfers/elterngeld.py [hunk 34]
```diff
@@ -1,56 +1,18 @@
...
 
 def elterngeld_m(
-    elterngeld_eink_relev_m: FloatSeries,
-    elternzeit_anspruch: BoolSeries,
-    elterngeld_eink_erlass_m: FloatSeries,
-    elterngeld_geschw_bonus_m: FloatSeries,
-    elterngeld_mehrlinge_bonus_m: FloatSeries,
+    elterngeld_eink_relev_m: float,
+    elternzeit_anspruch: bool,
+    elterngeld_eink_erlass_m: float,
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply gettsim/transfers/elterngeld.py:11]: also no idea. 

Would leave it as is for now.

File: gettsim/transfers/grunds_im_alter.py [hunk 35]
```diff
@@ -158,20 +138,20 @@ def grunds_im_alter_eink_m(
...
-        - tu_id.replace((eink_st_tu / anz_erwachsene_tu) / 12)
-        - tu_id.replace((soli_st_tu / anz_erwachsene_tu) / 12)
+        - (eink_st_tu / anz_erwachsene_tu / 12)
+        - (soli_st_tu / anz_erwachsene_tu / 12)
         - sozialv_beitr_gesamt_m
-    ).clip(lower=0)
+    )
 
-    return out
+    return max(out, 0.0)
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply gettsim/transfers/grunds_im_alter.py:146]: Why remove .0? 

Vectorization leads to problems if some results are float and others int.

[the remark this answers is not in the document]
Comment (hmgaudecker) [reply gettsim/transfers/grunds_im_alter.py:146]: Just readability / consistency. I did not realise that.

File: gettsim/transfers/kindergeld.py (hunk 17 above)
Comment (hmgaudecker) [reply gettsim/transfers/kindergeld.py]: > Why do you think the former makes more sense?

Because I would interpret "höchstalter" as "last age still eligible", not "first age ineligible". Given what you write, we should change to 24 in yaml then.

File: gettsim/transfers/kinderzuschl/kinderzuschl.py (hunk 19 above)
Comment (ChristianZimpelmann) [reply gettsim/transfers/kinderzuschl/kinderzuschl.py:48]: I think the latter.. I have made a note

File: gettsim/taxes/zu_verst_eink/eink.py (hunk 10 above)
Comment (ChristianZimpelmann) [reply gettsim/taxes/zu_verst_eink/eink.py:103]: Now I am confused. I thought you prefer `eink_vermiet`. I don't have a strong preference here.

File: gettsim/transfers/kindergeld.py (hunk 17 above)
Comment (ChristianZimpelmann) [reply gettsim/transfers/kindergeld.py]: > Because I would interpret "höchstalter" as "last age still eligible", not "first age ineligible".

Makes sense.

On the other hand, the numbers mentioned in the law are 18 and 25. I tend towards leaving the numbers as is and improve the parameter description.

Alternative would be to change numbers to 17 and 24 and adjust code accordingly.

**Comment by hmgaudecker** [comment]
On the phone and cannot directly reply at the right spots:

- I'd stick to vermietung and also vor_vermög_check (or the like) — simply
not reason to rename in this PR. Sorry, I see why that was confusing.
-17/24 + höchstalter, unless you come up with a really good name (I did
not, but that does not mean anything)

Thanks!

**Comment by ChristianZimpelmann** [comment]
> I'd stick to vermietung and also vor_vermög_check (or the like) — simply
not reason to rename in this PR. Sorry, I see why that was confusing.

I am still not sure what your prefered name is. We have two variables:
- input variable on monthly basis: current version and version before this PR: `eink_vermiet_m`
- defined function which calculates the value on yearly basis: before this PR `eink_vermietung`, current version: `eink_vermiet`

Just let me know how you would like to change it.

-17/24 + höchstalter, unless you come up with a really good name (I did
not, but that does not mean anything)

How about `altersgrenze`? I still think we should, if possible, use the parameter values mentioned in the law.

- Otherwise, all comments should be addressed now.
- I haven't implemented the user-provided aggregation-dictionary, yet. Should we postpone?

**Comment by hmgaudecker** [comment]
> > I'd stick to vermietung and also vor_vermög_check (or the like) — simply
> > not reason to rename in this PR. Sorry, I see why that was confusing.
> 
> I am still not sure what your prefered name is. We have two variables:
> 
>     input variable on monthly basis: current version and version before this PR: `eink_vermiet_m`
> 
>     defined function which calculates the value on yearly basis: before this PR `eink_vermietung`, current version: `eink_vermiet`

Ahh, I did not remember the former. Then we forgot to rename that, I believe. `eink_vermietung_m` / `eink_vermietung`, if that is not too much of a hassle

> How about `altersgrenze`? I still think we should, if possible, use the parameter values mentioned in the law.

Very good.

>     Otherwise, all comments should be addressed now.

Perfect! 
> 
>     I haven't implemented the user-provided aggregation-dictionary, yet. Should we postpone?

Sure, just open an issue. Or a PR directly, should be just a couple of hours' work, right? Then would be nice to have in release.

>     I tried to add python 3.10, but didn't get to run a working conda environment, yet. So let's postpone this for a bit.

Sure.
